### ISSUE TITLE
Create JUnit matchers for CompletionStage, Future, Optional

### ISSUE THREAD

**Issue by richdougherty** [playframework/playframework#5546]
It would help with writing Java tests to have some [matchers](https://github.com/junit-team/junit/wiki/Matchers-and-assertthat) for Java 8 types. I don't think they're in JUnit or Hamcrest [yet](https://github.com/hamcrest/JavaHamcrest/issues/82).

E.g.
```java
assertThat(completionStage, futureValue(is(3)));
assertThat(future, futureValue(is(3)));
assertThat(optional, isEmpty());
assertThat(optional, hasValue("x"));
```

**Comment by marcospereira** [comment]
This can be done as a separated library, I think. Or we can suggest that users adopt [AssertJ](http://joel-costigliola.github.io/assertj/) which has assertions for [CompletableFuture](http://joel-costigliola.github.io/assertj/core-8/api/org/assertj/core/api/Assertions.html#assertThat-java.util.concurrent.CompletableFuture-), [Optional](http://joel-costigliola.github.io/assertj/core-8/api/org/assertj/core/api/Assertions.html#assertThat-java.util.Optional-), [Streams](http://joel-costigliola.github.io/assertj/core-8/api/org/assertj/core/api/Assertions.html#assertThat-java.util.stream.Stream-) and many other types introduced in Java 8.

**Comment by richdougherty** [comment]
Yes it could be a separate library. The AssertJ library looks great. Is it compatible with JUnit/Hamcrest matchers or is it different?

**Comment by marcospereira** [comment]
@richdougherty it is compatible with JUnit and it has features that are equivalent to what Hamcrest offers.

**Comment by wsargent** [comment]
+1 for AssertJ

**Comment by npathai** [comment]
Matchers for Optional are available through [hamcrest-optional](https://github.com/npathai/hamcrest-optional). I am working towards other types.

```
<dependency>
  <groupId>com.github.npathai</groupId>
  <artifactId>hamcrest-optional</artifactId>
  <version>1.0</version>
</dependency>
```

**Comment by wsargent** [comment]
It may be easier to wait for Junit 5 at this point: http://junit.org/junit5/

**Comment by garretwilson** [comment]
I'm getting ready to use @npathai's implementation.  But someone mentioned JUnit 5. Are there indeed plans to address this in JUnit 5?

**Comment by lousanna** [comment]
The assertThat() function has been removed with JUnit 5:
http://junit.org/junit5/docs/current/user-guide/#writing-tests-assertions-third-party

**Comment by deepakab03** [comment]
So I am a bit confused: is there going to be optional support in Hamcrest or not? Can I contribute some code to this?

**Comment by gmethvin** [comment]
It appears Spotify has released a library effectively providing the helpers suggested here: https://github.com/spotify/java-hamcrest. Maybe we should just update the docs to point to that library?

**Comment by npathai** [comment]
@deepakab03 Hamcrest is still on 1.6 java and extensions for Java 8 are published separately. In future hamcrest may support java 8. I am author of [Hamcrest Optional](https://github.com/npathai/hamcrest-optional)

**Comment by mkurz** [comment]
Also seee #5546, let's see if we can upgrade to JUnit 5 soon...
